eventuals/catch.h

```diff
@@ -12,22 +13,38 @@ namespace eventuals {
 ////////////////////////////////////////////////////////////////////////
 
 struct _Catch {
-  template <typename K_, typename F_>
-  struct Continuation {
-    template <typename... Args>
-    void Start(Args&&... args) {
-      k_.Start(std::forward<Args>(args)...);
-    }
+  template <typename Error_, typename Handler_>
+  struct CatchHandler {
+    using error_type = Error_;
+    Handler_ handler_;
+  };
 
-    template <typename... Args>
-    void Fail(Args&&... args) {
-      using E = decltype(f_(std::forward<Args>(args)...)
+  ////////////////////////////////////////////////////////////////////////
+
+  template <typename... Ts>
+  struct IsSame {
+    static constexpr bool value = false;
+  };
+
+  template <typename Handler, typename Error>
+  struct IsSame<Handler, Error> {
+    static constexpr bool value =
+        std::is_same<typename Handler::error_type, Error>::value;
+  };
+
+  ////////////////////////////////////////////////////////////////////////
+
+  template <typename K_, bool AllSpecified_, typename... CatchHandlers_>
+  struct Continuation {
+    template <typename F, typename... Args>
+    void BuildAndInvoke(F&& f, Args&&... args) {
+      using E = decltype(f(std::forward<Args>(args)...)
                              .template k<void>(_Then::Adaptor<K_>{k_}));
 
       // TODO(benh): propagate eventual errors so we don't need to
       // allocate on the heap in order to type erase.
       e_ = std::unique_ptr<void, Callback<void*>>(
-          new E(f_(std::forward<Args>(args)...)
+          new E(f(std::forward<Args>(args)...)
```

Comment:
We actually now have enough information that we don't need to do dynamic memory allocation _except_ in the all case! Let's chat synchronously about what this can look like.